libraries/AP_AccelCal/AccelCalibrator.cpp

```diff
@@ -0,0 +1,464 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+// Authored by Jonathan Challinger, 3D Robotics Inc.
+
+#include "AccelCalibrator.h"
+#include <stdio.h>
+#include <AP_HAL/AP_HAL.h>
+
+const extern AP_HAL::HAL& hal;
+
+/*
+ * TODO
+ * - time out when not receiving samples
+ */
+
+////////////////////////////////////////////////////////////
+///////////////////// PUBLIC INTERFACE /////////////////////
+////////////////////////////////////////////////////////////
+
+AccelCalibrator::AccelCalibrator() :
+_conf_tolerance(ACCEL_CAL_TOLERANCE),
+_sample_buffer(NULL),
+_param_struct(*reinterpret_cast<struct param_t *>(&_param_array))
+{
+    clear();
+}
+/*
+    Select options, initialise variables and initiate accel calibration
+    Options:
+    Fit Type:       Will assume that if accelerometer static samples around all possible orientatio
+                    are spread in space will cover a surface of AXIS_ALIGNED_ELLIPSOID or any general 
+                    ELLIPSOID as chosen
+
+    Num Samples:    Number of samples user should will be gathering, please note that with type of surface
+                    chosen the minimum number of samples required will vary, for instance in the case of AXIS
+                    ALIGNED ELLIPSOID atleast 6 distinct samples are required while for generic ELLIPSOIDAL fit
+                    which will include calculation of offdiagonal parameters too requires atleast 8 parameters
+                    to form a distinct ELLIPSOID
+
+    Sample Time:    Time over which the samples will be gathered and averaged to add to a single sample for least
+                    square regression
+
+    offset,diag,offdiag: initial parameter values for LSQ calculation
+*/
+void AccelCalibrator::start(enum accel_cal_fit_type_t fit_type, uint8_t num_samples, float sample_time) {
+    start(fit_type, num_samples, sample_time, Vector3f(0,0,0), Vector3f(1,1,1), Vector3f(0,0,0));
+}
+
+void AccelCalibrator::start(enum accel_cal_fit_type_t fit_type, uint8_t num_samples, float sample_time, Vector3f offset, Vector3f diag, Vector3f offdiag) {
```

Comment:
why do we have a fit_type? Is there any purpose in something other than the 6 point fit?